Add resource pinning to Connect

gzdunek opened the pull request:
This is the frontend part of https://github.com/gravitational/teleport/pull/35139

e counterpart https://github.com/gravitational/teleport.e/pull/2810

We decided to show the `UnifiedResources` component as soon as possible, before the user preferences are fetched.
During that time, when we don't know yet which tab and view should be highlighted, nothing is selected (it also means that we don't show any skeleton).
To make the component work with such async behavior,  the `unifiedResourcePreferences` prop has been changed to accept `Attempt`.
We don't cache any preferences - if you open a new tab, they will be download again. By doing this, we will always have the fresh data.
The disadvantage of this solution is that we force the user to wait longer for the resources to be displayed. To remedy this, we may store the last `unifiedResourcePreferences` response in the workspace state and use it as a fallback, until we get the fresh data. 
I will add this after we make sure that this is the direction that we want to go.

gzdunek committed:
Expose tsh client in `AppContext`

gzdunek committed:
Change `unifiedResourcePreferences` to accept `Attempt`

gzdunek committed:
Add `unifiedResourcePreferencesFallback` and `updateUnifiedResourcesPreferencesAttempt` to `UnifiedResources`

gzdunek committed:
Add `UnifiedResourcePreferences` to workspace state, so it can be used as a fallback

gzdunek committed:
Add support for user preferences in Connect

gzdunek commented:
Okay, this is ready for review now.
I added a `useUserPreferences` hook that keeps all the logic related to the user preferences in unified resources in a single place.
As I wrote above, I also wanted to add a fallback for _unified resource preferences_ so the user sees the resources loading view faster, while the preferences are still fetched from the server. Having such fallback provides better UX, but it also raises a problem: the user may want to change the view, until the 'preferences from server' arrive. 
I wanted to handle that scenario nicely, so when it happens, we wait for the updated response to arrive before showing the resources.

avatus commented on web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:
```diff
@@ -428,9 +452,12 @@ export function UnifiedResources(props: UnifiedResourcesProps) {
-                  ? tab.value === DefaultTab.DEFAULT_TAB_PINNED
-                  : tab.value === DefaultTab.DEFAULT_TAB_ALL
+                unifiedResourcePreferences
+                  ? // TODO(gzdunek): selected tab should be taken from unifiedResourcePreferences
```
it looks like you're doing this TODO in the change? I might be confused

avatus commented on web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx:
```diff
@@ -190,14 +205,55 @@ function Resources(props: {
   }
 
+  function getPinning(): UnifiedResourcesPinning {
+    // optimistically assume that pinning is supported
```
probably a good idea as there isn't a way to "disable" pinning currently, and is only as a way to support old clusters. but since this is coming out in 15(?), the n-1 versions should all support pinning

gzdunek replied:
This comes out in v15, but pinning has been added in v14.1.5. We still have to support v14.0.0.

ravicious reviewed (commented):
I didn't have time to get to the last commit (which is the most important one), I'll continue the review tomorrow.

gzdunek committed:
Add more `UnifiedResources` stories

gzdunek committed:
Add descriptions to error messages

gzdunek committed:
Simplify errors rendering

gzdunek committed:
Use smaller margins in `DocumentCluster`

gzdunek committed:
Parse unified resource preferences read from disk in zod

gzdunek commented on web/packages/shared/components/UnifiedResources/UnifiedResources.story.tsx:
```diff
@@ -126,7 +128,7 @@ const story = ({
         setParams={() => undefined}
         pinning={pinning}
-        unifiedResourcePreferences={userPrefs}
+        unifiedResourcePreferencesAttempt={makeSuccessAttempt(userPrefs)}
```
Added (also for updating).

ravicious commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
```diff
@@ -0,0 +1,189 @@
+  async function updateUserPreferences(
+    newPreferences: Partial<UserPreferences>
+  ): Promise<void> {
+    // If we don't have the preferences fetched yet,
+    // we start from updating the workspace (so the user sees an update in the UI).
+    //
+    // Since we know that the preferences have changed,
+    // we don't want to show the user the preferences from the "get"
+    // request that are outdated.
+    // Instead, we replace a request in that attempt with an "update" request,
+    // falling back to "get" response, only if the update failed.
+    // Thanks to that, the user sees a continuous loading screen.
```
The code which achieves that is a bit hard to understand, mostly because it uses some approaches which are not common in our codebase (`useAsync((value) => value)`, storing a promise in a ref).

The way I understand this problem is that if someone clicks a button to change from cards to a list but we're yet to know what the actual preferences are like, we want to disregard the actual prefs we get from the cluster and wait for the updated prefs instead.

Isn't this a matter of having two different attempts, one that fetches the current prefs and another that updates them, and returning only one of them from the hook? If an update is in progress, we take the prefs from the update response. If there's no update, we use the actual prefs.

ravicious commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
```diff
@@ -0,0 +1,189 @@
+        userPreferences: newPreferences,
+      })
+    );
+  appContext.workspacesService.useState();
```
This has some heavy implications when it comes to performance, do we absolutely need it here? I feel like this should be reserved only to components which are not bound to a particular workspace.

gzdunek replied:
Yeah, I think we can remove it.

ravicious commented on web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:
If the request to get prefs takes longer than the request to get resources, we render the resources and then suddenly change their layout when the prefs arrive. I don't think it should work this way. TBC because I have spotty internet

https://github.com/gravitational/teleport/assets/27113/a9695b04-af62-41a8-8676-8d7b98f6b933

gzdunek replied:
Sorry, in the end I didn't add it, fixed.

gzdunek committed:
Fix DELETE IN comments

gzdunek committed:
Move `unifiedResourcePreferencesSchema` outside the function

gzdunek committed:
Remove `workspacesService.useState()`

gzdunek committed:
Show skeleton when `unifiedResourcePreferencesAttempt` is processing

gzdunek committed:
Run Prettier

gzdunek commented:
Eh, now I thought about an edge case related to the fact that we return the updated preferences from the server.

If you are on a cluster that supports resource pinning but doesn't support the list view mode, changing the view to list and then pinning a resource will reset the view, because the server will return `VIEW_MODE_UNSPECIFIED`. 

I think we could fix this by removing the default values from `useUserPreferences` and handle that in `UnifiedResources`:
* if the unified view preferences object is empty or `viewMode === VIEW_MODE_UNSPECIFIED`  lock the resource view icon
* if the unified view preferences object is empty or `defaultTab === DEFAULT_TAB_UNSPECIFIED`  lock the pinning tab

gzdunek committed:
Import `renderHook` from correct package

gzdunek committed:
Add `retryWithRelogin`

ravicious commented:
Re concurrent requests, wouldn't it be simpler to add support for `AbortSignal` to `retryWithRelogin`?

ravicious commented on web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:
```diff
@@ -482,7 +482,8 @@ export function UnifiedResources(props: UnifiedResourcesProps) {
               resourcesFetchAttempt.status === 'processing' ||
-              getPinnedResourcesAttempt.status === 'processing'
+              getPinnedResourcesAttempt.status === 'processing' ||
+              unifiedResourcePreferencesAttempt.status === 'processing'
```
I don't think it's enough, you'll end up displaying both the resources and the loading skeleton in a situation where the resources have been fetched but we're still to receive prefs.

ravicious replied:
In my local version, I got around this by doing this:

```typescript
            mappedResources={
              // Hide the resources until the preferences are fetched.
              // ViewComponent supports infinite scroll, so it shows both already loaded resources
              // and a loading indicator if needed.
              hasFinished(unifiedResourcePreferencesAttempt)
                ? resources.map(unifiedResource => ({
                    item: mapResourceToViewItem(unifiedResource),
                    key: generateUnifiedResourceKey(unifiedResource.resource),
                  }))
                : []
            }
```

gzdunek replied:
Thanks, done.

ravicious commented on web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:
```diff
@@ -482,7 +482,8 @@ export function UnifiedResources(props: UnifiedResourcesProps) {
               resourcesFetchAttempt.status === 'processing' ||
-              getPinnedResourcesAttempt.status === 'processing'
+              getPinnedResourcesAttempt.status === 'processing' ||
+              unifiedResourcePreferencesAttempt.status === 'processing'
```
I'm not 100% sure, but those checks should probably check `''` as well to show the loading skeleton ASAP. Otherwise they'll wait for `useEffect` or something else to start fetching and change the state from `''` to `'processing'`. Unless the hook for fetching the resources has some extra logic about turning the attempt status back to `''` when it's done, idk.

gzdunek replied:
Yeah, `useKeyBasedPagination` has that logic unfortunately, but I added it for other cases.

gzdunek committed:
Revert sync `unifiedResourcePreferences`

gzdunek committed:
Refactor `useUserPreferences`

gzdunek committed:
Rename `userPreferences.test.tsx` to `useUserPreferences.ts`

gzdunek committed:
Do not show 'pinning not supported' and 'no results' errors at the same time

We know upfront that we won't fetch resources,
so we can avoid showing errors related to them.

gzdunek committed:
Check for `''` too when calculating `isProcessing`

gzdunek committed:
Small fixes

gzdunek committed:
Remove an unneeded story

gzdunek committed:
Fix tests and lint

ravicious reviewed (commented):
This seems to work great and the code is much easier to understand now. However, the user prefs in the Web UI do not live in a state that's local to each tab, that assumption was wrong, so I wonder what to do with it in Connect. See my last comment.

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
Ah, I completely forgot that Web UI shares the preferences updates through `localStorage`.

Probably we could do it as you said, but OTOH I asked Kenny about it some time ago (although I didn't know then that it works in that way in Web UI): 
>This should update the global cluster preferences (on the backend side) but on the UI side it should apply only to the active tab and new tabs IMO. Do you agree?

https://gravitational.slack.com/archives/C03FJA391M3/p1700579896052849

So maybe it's fine to keep the current behavior?

gzdunek committed:
Improve handling errors when initial request is superseded, notify about preferences update failure

gzdunek committed:
Improve comments

gzdunek committed:
Revert `unifiedResourcePreferences` to be required (as it was originally)

gzdunek committed:
Improve condition for initial fetch

gzdunek committed:
Test if the stale response is discarded

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
```diff
@@ -0,0 +1,197 @@
+        initialFetchAttemptAbortController.current.abort();
+      }
+
+      const [prefs, error] = await runUpdateAttempt(newPreferences);
```
Right, added a warning.

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.test.tsx:
```diff
@@ -0,0 +1,223 @@
+    });
+  });
+
+  test('when the request for fetching preferences is in-flight', async () => {
```
Actually, I wanted to make the full test name from `describe` and `test`, so now (after some corrections) it is:
`updating preferences works correctly when the initial preferences have not been fetched yet`

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.test.tsx:
```diff
@@ -0,0 +1,223 @@
+    act(() => resolveGetUserPreferencesPromise(null));
+    await act(() => getUserPreferencesPromise);
+    act(() => resolveUpdateUserPreferencesPromise(null));
+    await act(() => updateUserPreferencesPromise);
```
Since the initial fetch request is aborted, I couldn't test exactly that (trying to resolve a rejected promise was throwing an error). 
I added checks that verify if the users's unified resource preferences are not overwritten when a different preferences are returned from `update`.

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
```diff
@@ -0,0 +1,197 @@
+    updateUnifiedResourcePreferencesStateAndWorkspace,
+  ]);
+
+  const updateUserPreferences = useCallback(
```
Yeah, that is handled by `useAsync` - when a new request is created, the previous one is rejected with `CancelledError`.

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
```diff
@@ -0,0 +1,197 @@
+
+      const hasUpdateSupersededInitialFetch =
+        initialFetchAttempt.status !== 'success' &&
+        supersededInitialFetchAttempt.status === '';
```
Agree, I think that changing the condition makes sense.

gzdunek committed:
Support `labelsViewMode`

gzdunek committed:
Remove `width` from expand/collapse button, so it looks good independently of font weight

gzdunek commented:
Added support for `labelsViewMode`.

gzdunek committed:
Run prettier

ravicious committed:
Expand comment regarding updating cluster prefs

ravicious commented on web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:
In the videos I added 5 seconds of sleep to the tshd handlers that deal with user prefs. The loading skeleton shifts many times over when opening a new tab and it's even worse when reopening the app. I suppose this is because the skeleton component randomizes the lengths each time it's rendered instead of only on the initial mount or something. Maybe this is something that could be worked around with the `key` prop?

I'd say this is something to adjust in a future PR. I assume it's a new regression. I haven't checked if this affects the Web UI as well.

Also, the app asks me if I want to reopen the tabs even though I have only the default cluster app opened. I guess this is related to the changes with query params in `DocumentCluster`.

https://github.com/gravitational/teleport/assets/27113/231f9a7d-a2fa-4894-94da-f90a00dc1f77


https://github.com/gravitational/teleport/assets/27113/d0795ab8-a191-484b-80b8-8a5f4c922903

gzdunek replied:
>I suppose this is because the skeleton component randomizes the lengths each time it's rendered instead of only on the initial mount or something.

Yeah, this is the reason. I think it was less visible in the Web UI because there the component doesn't re-render skeleton that frequently.

I will try to fix it tomorrow., maybe adding `key` will be enough.

gzdunek replied:
>Also, the app asks me if I want to reopen the tabs even though I have only the default cluster app opened. I guess this is related to the changes with query params in DocumentCluster.

Hmm, I think so. It doesn't happen to me if the query params are the same.

gzdunek committed:
Improve test names

gzdunek committed:
Fix grammar

gzdunek committed:
Add a comment about storing preferences in the document

gzdunek commented on web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts:
```diff
@@ -145,7 +150,10 @@ export function useUserPreferences(clusterUri: ClusterUri): {
+        // We always try to update cluster preferences,
+        // so the user sees the recent pinned resources.
+        // We don't do it for unified resources preferences
+        // because we don't want to suddenly change the view.
```
It's better now, thanks :)

gzdunek committed:
Run prettier

gzdunek committed:
Fix license
